examples/remote/binary_dependencies/cargo/crates.bzl

```diff
@@ -9,6 +9,144 @@ load("@bazel_tools//tools/build_defs/repo:git.bzl", "new_git_repository")  # bui
+        fail("An invalid argument has been provided. Please pass a crate name or a list of crate names")
+
+    if not dependencies:
+        return []
+
+    if type(deps) == "string":
+        deps = [deps]
+
+    errors = []
+    crates = []
+    for crate in deps:
+        if crate not in dependencies[native.package_name()]:
+            errors.append(crate)
+        else:
+            crates.append(dependencies[native.package_name()][crate])
+
+    if errors:
+        fail("Missing crates `{}` for package `{}`. Available crates `{}".format(
+            errors,
+            native.package_name(),
+            dependencies[native.package_name()],
+        ))
+
+    return crates
+
+def all_crates(normal = False, proc_macro = False, dev = False, dev_only = False):
+    """EXPERIMENTAL -- MAY CHANGE AT ANY TIME: Finds the fully qualified label of all requested direct crate dependencies \
+    for the package where this macro is called.
+
+    If no parameters are set, all normal and proc_macro dependencies are returned.
+    Setting any one flag will otherwise impact the contents of the returned list
+
+    WARNING: This macro is part of an expeirmental API and is subject to change.
+
+    Args:
+        normal (bool, optional): If True, normal dependencies are included in the
+            output list. Defaults to False.
+        proc_macro (bool, optional): If True, proc_macro dependencies will be
+            included in the output list. Defaults to False.
+        dev (bool, optional): If True, dev dependencies are included when the
+            `normal` and `proc_macro` parameters are used. Defaults to False.
+        dev_only (bool, optional): If True, only development dependencies will be
+            returned by this list. This paramter otherwise follows the same rules
+            as `dev`. Defaults to False.
+
+    Returns:
+        list: A list of labels to cargo-raze generated targets (str)
+    """
+
+    # Determine the relevant maps to use
+    dependencies = dict()
+    all_maps = []
+    if normal:
+        if not dev_only:
+            all_maps.append(_DEPENDENCIES)
+        if dev or dev_only:
+            all_maps.append(_DEV_DEPENDENCIES)
+    if proc_macro:
+        if not dev_only:
+            all_maps.append(_PROC_MACRO_DEPENDENCIES)
+        if dev or dev_only:
+            all_maps.append(_DEV_PROC_MACRO_DEPENDENCIES)
+
+    # Default to always using normal dependencies
+    if not all_maps:
+        if not dev_only:
+            all_maps.append(_DEPENDENCIES)
+        if dev or dev_only:
+            all_maps.append(_DEV_DEPENDENCIES)
+
+    if not all_maps:
+        fail("Failed to add at least 1 map to the `all_maps` list with parameters: " +
+             "normal = {normal}, proc_macro = {proc_macro}, dev = {dev}, dev_only = {dev_only}".format(
+                 normal = normal,
+                 proc_macro = proc_macro,
+                 dev = dev,
+                 dev_only = dev_only,
+             ))
```

Comment:
This map setup stuff is jank. I'm hoping that we can clean this up by updating the semantics of the `dev` argument.

Reply:
The use cases this needs to be able to handle is,
1. Grab all normal dependencies
2. Grab all proc_macro dependencies
3. Grap all normal dev dependencies
4. Grab all proc_macro dev dependencies
5. Grab dev and default normal dependencies
6. Grab dev and default proc_macro dependencies

What would clean this up best and solve for those? Or do you think that list is not quite accurate?

Reply:
How about

`normal, normal_dev, proc_macro, proc_macro_dev`?

It's one more argument, but I think the semantics are better.

Reply:
I think that's good, I'll update that

Reply:
Updated